### PR TITLE
Add batching to `IterableDataset`

### PR DESCRIPTION
**Pull request by lappemic**
I've taken a try at implementing a batched `IterableDataset` as requested in issue #6279. This PR adds a new `BatchedExamplesIterable` class and a `.batch()` method to the `IterableDataset` class.

The main changes are:
1. A new `BatchedExamplesIterable` that groups examples into batches.
2. A `.batch()` method for `IterableDataset` to easily create batched versions.
3. Support for shuffling and sharding to work with PyTorch DataLoader and multiple workers.

I'm not sure if this is exactly what you had in mind and also have not fully tested it atm, so I'd really appreciate your feedback. Does this seem like it's heading in the right direction? I'm happy to make any changes or explore different approaches if needed.

Pinging @lhoestq

### CODE REVIEW

**Comment by lhoestq** [comment]
Cool ! Thanks for diving into it :)

Your implementation is great for and indeed supports shuffling and batching, you just need to additionally account for state_dict (for dataset [checkpointing+resuming](https://huggingface.co/docs/datasets/main/en/use_with_pytorch#checkpoint-and-resume))

That being said, I believe the implementation can be made simpler by relying on `IterableDataset.map()` which already implements all this. Maybe something like

```python

def batch(self, batch_size: int, drop_last_batch: bool = False) -> "IterableDataset":
    def batch(unbatched: dict[str, list]) -> dict[str, list]:
        return {k: [v] for k, v in unbatched}

    return self.map(batch, batched=True, batch_size=batch_size, drop_last_batch=drop_last_batch)
```

And this way no need to reimplement everything !

(my only small concern is that it's not an Arrow-optimized function so it requires the examples to be manipulated as python objects even if the original data is in Arrow format (e.g. when streaming Parquet files) but it's not a big deal and we can see later if we need to optimize this)

**Comment by lappemic** [comment]
Thanks a lot for the feedback @lhoestq! I definitely could have saved some time looking into it properly first. 😅 

Implemented the `.batch()` method, added a proper docsrtring for documentation, and added tests.

Let me know what you think and if this needs some update.

File: src/datasets/iterable_dataset.py [hunk 1]
```diff
@@ -2885,6 +2885,28 @@ def _resolve_features(self):
...
 
+    def batch(self, batch_size: int, drop_last_batch: bool = False) -> "IterableDataset":
+        """
+        Group samples from the dataset into batches.
+
+        Args:
+            batch_size (int):
+                The number of samples in each batch.
+            drop_last_batch (bool, defaults to `False`):
+                Whether to drop the last incomplete batch.
```
Comment (lhoestq) [inline src/datasets/iterable_dataset.py:2896]:
```suggestion
            batch_size (`int`): The number of samples in each batch.
            drop_last_batch (`bool`, defaults to `False`): Whether to drop the last incomplete batch.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 2]
```diff
@@ -205,6 +205,43 @@ for epoch in range(n_epochs):
...
+When set to `True`, it will discard the last batch if it's smaller than the specified `batch_size`. 
+This can be useful in scenarios where your downstream processing requires all batches to be of the same size:
+
+```python
+batched_dataset = dataset.batch(batch_size=32, drop_last_batch=True)
+```
+
+Batching with `IterableDataset` combines the benefits of processing data in larger chunks with the memory efficiency of streaming data. 
+This makes it an excellent choice for working with large datasets that don't fit in memory, especially when used with deep learning frameworks that expect batched inputs.
+
```
Comment (lhoestq) [inline docs/source/about_mapstyle_vs_iterable.mdx:244]: feels a bit redundant

```suggestion
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 3]
```diff
@@ -205,6 +205,43 @@ for epoch in range(n_epochs):
         pass
 ```
 
+## Batching data in IterableDataset
```
Comment (lhoestq) [inline docs/source/about_mapstyle_vs_iterable.mdx:208]: this can go in stream.mdx instead ?

Btw in the future it would be cool to have `Dataset.batch()` as well :)

**Review by lhoestq — COMMENTED** [review]
Nice thanks ! I added a few suggestions before we merge.

**Comment by lappemic** [comment]
Thanks for the feedbak @lhoestq!

Applied it and referenced the `batched=True` option in the `map` function and highlighted the difference. Hope i got this right.

**Review by lhoestq — APPROVED** [review]
lgtm !
